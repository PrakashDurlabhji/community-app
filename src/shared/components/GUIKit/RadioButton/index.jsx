/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
/**
 * Radio button component.
 */
import React, { useRef, useState } from 'react';
import PT from 'prop-types';
import _ from 'lodash';
import './style.scss';

import { config } from 'topcoder-react-utils';

function RadioButton({
  options, onChange, size, errorMsg,
}) {
  const [internalOptions, setInternalOptions] = useState(options);
  const optionsWithKey = internalOptions.map((o, oIndex) => ({ ...o, key: oIndex }));
  let sizeStyle = size === 'lg' ? 'lgSize' : null;
  if (!sizeStyle) {
    sizeStyle = size === 'xs' ? 'xsSize' : 'smSize';
  }
  const delayedOnChange = useRef(
    _.debounce((q, cb) => cb(q), config.GUIKIT.DEBOUNCE_ON_CHANGE_TIME),
  ).current;

  return (
    <React.Fragment>
      <div className="radioButtonContainer" styleName={`radioButtonContainer ${sizeStyle}`}>
        {optionsWithKey.map(o => (
          <div key={o.key} styleName="radioButton" className="radioButton">
            <label styleName="container">
              <input
                type="radio"
                checked={o.value}
                onChange={() => {
                  const newOptions = optionsWithKey.map(oWithKeyTmp => ({
                    label: oWithKeyTmp.label,
                    value: o.key === oWithKeyTmp.key,
                  }));
                  setInternalOptions(newOptions);
                  delayedOnChange(_.cloneDeep(newOptions), onChange);
                }}
              />
              <span styleName={`checkmark ${errorMsg ? 'hasError' : ''}`} />
            </label>
            {o.label ? (<span styleName="label">{o.label}</span>) : null}
          </div>
        ))}
      </div>
      {errorMsg ? (<span styleName="errorMessage">{errorMsg}</span>) : null}
    </React.Fragment>
  );
}

RadioButton.defaultProps = {
  onChange: () => {},
  size: 'sm',
  errorMsg: '',
};

RadioButton.propTypes = {
  options: PT.arrayOf(
    PT.shape({
      label: PT.string,
      value: PT.bool.isRequired,
    }),
  ).isRequired,
  onChange: PT.func,
  size: PT.oneOf(['xs', 'sm', 'lg']),
  errorMsg: PT.string,
};

export default RadioButton;
