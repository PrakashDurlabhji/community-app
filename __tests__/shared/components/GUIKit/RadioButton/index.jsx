import React from 'react';
import Renderer from 'react-test-renderer/shallow';

import RadioButton from 'components/GUIKit/RadioButton';


const rnd = new Renderer();

it('Default render', () => {
  rnd.render((<RadioButton options={[{ label: 'radio', value: true }]} />));
  expect(rnd.getRenderOutput()).toMatchSnapshot();
});
