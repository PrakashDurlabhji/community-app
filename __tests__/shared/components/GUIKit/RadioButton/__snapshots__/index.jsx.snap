// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default render 1`] = `
<React.Fragment>
  <div
    className="radioButtonContainer src-shared-components-GUIKit-RadioButton-___style__radioButtonContainer___3YAY0 src-shared-components-GUIKit-RadioButton-___style__smSize____lb3s"
  >
    <div
      className="radioButton src-shared-components-GUIKit-RadioButton-___style__radioButton___2Jtbj"
    >
      <label
        className="src-shared-components-GUIKit-RadioButton-___style__container___3nVSA"
      >
        <input
          checked={true}
          onChange={[Function]}
          type="radio"
        />
        <span
          className="src-shared-components-GUIKit-RadioButton-___style__checkmark___1iCFJ"
        />
      </label>
      <span
        className="src-shared-components-GUIKit-RadioButton-___style__label___MCxnk"
      >
        radio
      </span>
    </div>
  </div>
</React.Fragment>
`;
